ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -271,4 +489,376 @@ private void updateMapHelper(Map<String, Long> map, String key, Long value) {
     Long newValue = map.containsKey(key) ? map.get(key) + value : value;
     map.put(key, newValue);
   }
+
+  /**
+   * Helper function for container buckets for blob expiration/deletion related updates.
+   * @param results the {@link ScanResults} to be updated
+   * @param indexValue the PUT {@link IndexValue} of the expiring/deleting blob
+   * @param expOrDelTimeInMs either the expiration or deletion time of the blob in ms
+   * @param operator indicating the operator of the update, 1 for add and -1 for subtract
+   */
+  private void handleContainerBucketUpdate(ScanResults results, IndexValue indexValue, long expOrDelTimeInMs,
+      int operator) {
+    if (isWithinRange(results.containerForecastStartTimeMs, results.containerLastBucketTimeMs, expOrDelTimeInMs)) {
+      results.updateContainerBucket(results.getContainerBucketKey(expOrDelTimeInMs),
+          String.valueOf(indexValue.getServiceId()), String.valueOf(indexValue.getContainerId()),
+          indexValue.getSize() * operator);
+    }
+  }
+
+  /**
+   * Helper function for log segment buckets for blob expiration/deletion related updates.
+   * @param results the {@link ScanResults} to be updated
+   * @param indexValue the PUT {@link IndexValue} of the expiring/deleting blob
+   * @param expOrDelTimeInMs either the expiration or deletion time of the blob in ms
+   * @param operator indicating the operator of the update, 1 for add and -1 for subtract
+   */
+  private void handleLogSegmentBucketUpdate(ScanResults results, IndexValue indexValue, long expOrDelTimeInMs,
+      int operator) {
+    if (isWithinRange(results.logSegmentForecastStartTimeMs, results.logSegmentLastBucketTimeMs, expOrDelTimeInMs)) {
+      results.updateLogSegmentBucket(results.getLogSegmentBucketKey(expOrDelTimeInMs), indexValue.getOffset().getName(),
+          indexValue.getSize() * operator);
+    }
+  }
+
+  /**
+   * Helper function to process new PUT entries and make appropriate updates to the given {@link ScanResults}.
+   * @param results the {@link ScanResults} to apply the updates to
+   * @param putValue the {@link IndexValue} of the new PUT
+   */
+  private void processNewPut(ScanResults results, IndexValue putValue) {
+    long expiresAtMs = putValue.getExpiresAtMs();
+    if (!isExpired(expiresAtMs, results.containerForecastStartTimeMs)) {
```

Comment:
this is to guard against the condition, that between the time new put entry was added to the queue, and the queueProcessor processes the entry, if the entry is expired?